### DISCUSSION TITLE
add IRIS (and other objects) in Project

### DISCUSSION THREAD

**Discussion by atravert** [spectrochempy/spectrochempy#429] [Developer discussions]
I am trying to add other types of objects in Project, for instance IRIS objects, which have several attributes including closely related NDDatasets, so I think it can make sense to include IRIS objects as such isnstead of the indivudual NDDatatsets (similar things could be done for PCA, MCR-ALS, etc...).

I have -- alsmost :-) -- succeeded in [feature/dev_project](https://github.com/atravert/spectrochempy/tree/feature/dev_project), except that I can't figure what to do for a proper serialization the IRIS object. `json_serialiser` raises the following error: "No encoding handler for data type <class 'method'>" -- I think it is trying to serialize the IRIS `__init__` method... which is a bit strange to me . 

Anyway, I clearly miss something :-)  There is no emergency on this, but if you have some time, here is a minimum code, working fine, excep for the last line !  
```
import spectrochempy as scp

# import a dataset, make iris
X = scp.read("irdata/CO@Mo_Al2O3.SPG")[:, 2250.0:1950.0]

pressures = [0.003, 0.004, 0.009, 0.014, 0.021, 0.026,
             0.036, 0.051, 0.093, 0.150, 0.203, 0.300,
             0.404, 0.503, 0.602, 0.702, 0.801, 0.905,
             1.004, ]

c_pressures = scp.Coord(pressures, title="pressure", units="torr")

X.y = c_pressures
iris = scp.IRIS(X, "langmuir", q=[-8, -1, 50])

# add the dataset and iris in a project

proj = scp.Project(name="proj")
proj.add_dataset(X)
proj.add_iris(iris)

print(proj)
proj.save(confirm=False)
```

**Comment by fernandezc** [comment]
Hi Arnaud,
did you define the __dir__() method in your IRIS object. I didn't check but if you did not, you must do it, specifying which are the attribute required to recreate the object. His is used for copy also. Look at Coord for instance.

**Comment by fernandezc** [comment]
I mean the special method `__dir__`

**Comment by fernandezc** [comment]
By the way you probably need also to define `__eq__` so that we know how to compare IRIS objects. This is what is done for Script for example.

**Comment by fernandezc** [comment]
I had a look to your code. Yes it is due to the lack of `__dir__`. 

I have a little problem with your implementation. I don't find it very coherent to refer to an object defined in `analysis` as if it was part of `core`. I suggest to create a metaclass `Projectable` for example, which would be subclassed by the objects to be added to a project: At least it should contain the `implements` method and the `__dir__` method.  This metaclass would reside in the core (maybe in the `dataset` directory? 
Then `IRIS` or any other object to put in the project, would be a subclass of `Projectable`:

```python
class Projectable(HasTraits):  
     def __dir__():
          raise NotImplementedError

     def implements():
          raise NotImplementedError

class IRIS(Projectable): 
     ....
     def __dir__():
          Etc.... 
```

this way in Project we should be able to make generic methods to add foreign objects:

en modifiant 
```python
    def __getattr__(self, item):

        if 
        .....

        elif item.startswith('add_'):
              klass = item.replace("add_","")
              return add_foreign( ...., klass=klass)
        else:
            raise AttributeError(
                "`%s` has no attribute `%s`" % (type(self).__name__, item)
            )

def add_foreign(self, obj, klass=None):
       etc.... 
```

Well, this is just an idea.

**Reply by fernandezc** [reply]
I forgot, this way we can call transparently:
```python
proj.add_iris(iris) 
```

**Reply by fernandezc** [reply]
This of course needs some work. I am not sure that my implementation of `__getattr__` is correct.

**Reply by fernandezc** [reply]
this seems more correct:
```python
elif item.startswith('add_'):
              return add_foreign
```

**Reply by fernandezc** [reply]
This way we can simplify Project, 
with only one type of add_xxx method, because for now they all looks similar.   But I did not investigate too much 🙂

**Comment by atravert** [comment]
wow, thank you very much ! I will try this clean implementation, probably  next week. 

For now it works, with save and load  -- for the loading I had also to change IRIS `__init__()` so that it generate an object with the `__dir__` attributes without positional argument. I have a last (?) issue: after loading the IRIS attributes of NDDataset type  (X, K, f,...) appear as dicts. Any clue as how to fix it neatly  ?  ( this arises in ndio.py, line 388. I have few ideas as how to  fix it...  inelegantly 😉 ...

```
import spectrochempy as scp
# import a dataset, make iris
X = scp.read("irdata/CO@Mo_Al2O3.SPG")[:, 2250.0:1950.0]

pressures = [0.003, 0.004, 0.009, 0.014, 0.021, 0.026,
             0.036, 0.051, 0.093, 0.150, 0.203, 0.300,
             0.404, 0.503, 0.602, 0.702, 0.801, 0.905,
             1.004, ]

c_pressures = scp.Coord(pressures, title="pressure", units="torr")

X.y = c_pressures
iris = scp.IRIS(X, "langmuir", q=[-8, -1, 50])

# add the dataset and iris in a project

proj = scp.Project(name="proj")
proj.add_dataset(X)
proj.add_iris(iris, name='iris')

#%%

proj.save(confirm=False)

proj2 = scp.Project().load("proj.pscp")

type(proj2['iris'].K)  # returns a 
```

**Reply by fernandezc** [reply]
OK, I'll try to look at this sometimes later :-).

**Reply by atravert** [reply]
I have solved this IRIS dict<->NDDataset issue in NDIO `load()` :

```
                    elif key in ["_irises"]:
                        irises = [item_to_attr(IRIS(), js) for js in val]
                        # iris f, K, X appear as dicts, we convert them
                        # back to datasets:
                        for key, val in irises[0].__dict__.items():
                            if key in ["f", "K", "X"]:
                                setattr(irises[0], key, item_to_attr(NDDataset(), val))
                        obj.irises = irises
```

It seems reasonable to me, but let me knwow :-)

**Comment by fernandezc** [comment]
Regarding the implementation, I'm pretty sure we can simplify this (without passing any parameters), but in case we need to pass the `klass` then the correct implementation would be something like this:

```python
from functools import partial

class Project(...

    def __getattr__(self, item):

        if 
        .....

        elif item.startswith('add_'):
              klass = item.replace("add_","")
              return partial(add_foreign, klass=klass)
        else:
            raise AttributeError(
                "`%s` has no attribute `%s`" % (type(self).__name__, item)
            )

   def add_foreign(self, obj, klass=None):
          etc.... 
```
Another way would be to define simply an `add` and. `del` method for everything and check the `implements`atribute to determine the rest of the processing. 
This of course need quite a significant rewriting of project, but with this we could be able to a dd any subclass of `Projectable`, including Dataset, and Script which already implements the required methods.  
I let you work on this!  it will be quite an improvement of this object, for sure.

**Reply by atravert** [reply]
OK. I will work on it, as I am not very comfortable with class inheritence and related issues, it will be a good exercise - hopefully doable for me...

**Reply by fernandezc** [reply]
Yes, probably it is important to try. It will allow you to make a big step forward in coding and probably simplify some tasks later. Ask if necessary!

**Comment by atravert** [comment]
I have started rewriting project so that any type of object can be added with - as suggested - the addtion of a `Projectable` class to deal with objects that have no names (e.g. ndarray, str,...). -- I have also added a date field because at the end  I would like being able to sort objects by name, type, date or size -- and a single add() and remove method 

Below is a sample script and output showing how it behaves when adding/removing objects. Let me know if this seems a good start or not !

In the next steps, I will look at
- save/load
- adding and running scripts
- sorting by name, type, date or size


```
import spectrochempy as scp
import numpy as np

A = scp.read("irdata/CO@Mo_Al2O3.SPG")
my_array = np.array([3.5, 2.5])

proj = scp.Project(name="top_project")
proj.add(A)
proj.add(my_array, name="my_array")
print(proj)

sub_proj = scp.Project(('sub_project'), objnames=['a string'], name="sub_proj")
proj.add(sub_proj)
print(proj)

proj.remove("my_array")
print(proj)

```

output:

```
SpectroChemPy's API - v.0.4.3.dev11+g9732da10c
© Copyright 2014-2022 - A.Travert & C.Fernandez @ LCS

top_project (Project):
   ⤷ CO@Mo_Al2O3 (NDDataset)
   ⤷ my_array (ndarray)

top_project (Project):
   ⤷ CO@Mo_Al2O3 (NDDataset)
   ⤷ my_array (ndarray)
   ⤷ sub_proj (Project):
      ⤷ a string (str)

top_project (Project):
   ⤷ CO@Mo_Al2O3 (NDDataset)
   ⤷ sub_proj (Project):
      ⤷ a string (str)


```
